src/ivory-px4-hw/SMACCMPilot/Hardware/LSM303D/SPI.hs

```diff
@@ -129,59 +141,51 @@ lsm303dSPISensorManager conf (BackpressureTransmit req_chan res_chan) init_chan
           emit req_e do_read_req
 
 convert_mag_sample :: Config
-                   -> Maybe MagCal
+                   -> MagCal
                    -> Ref s1 ('Struct "spi_transaction_result")
                    -> Ref s2 ('Struct "magnetometer_sample")
                    -> Ivory eff ()
-convert_mag_sample c Nothing res s =
-  convert_sample (magSensitivityGauss c) res (s ~> M.sample)
-convert_mag_sample c (Just MagCal {..}) res s =
-  convert_sample_cal
-    mag_cal_x_offset mag_cal_y_offset mag_cal_z_offset
-    (magSensitivityGauss c)
-    res
-    (s ~> M.sample)
+convert_mag_sample c (MagCal xyz_cal) res s =
+  convert_sample xyz_cal (magSensitivityGauss c) res (s ~> M.sample)
 
 convert_acc_sample :: Config
+                   -> AccelCal
                    -> Ref s1 ('Struct "spi_transaction_result")
                    -> Ref s2 ('Struct "accelerometer_sample")
                    -> Ivory eff ()
-convert_acc_sample c res s = convert_sample (accelSensitivityMSS c) res (s ~> A.sample)
+convert_acc_sample c (AccelCal xyz_cal) res s =
+  convert_sample xyz_cal (accelSensitivityMSS c) res (s ~> A.sample)
 
-convert_sample :: IFloat
+convert_sample :: XyzCal
+               -> IFloat
                -> Ref s1 ('Struct "spi_transaction_result")
                -> Ref s2 ('Struct "xyz")
                -> Ivory eff ()
-convert_sample = convert_sample_cal 0 0 0
-
-convert_sample_cal :: Sint16 -> Sint16 -> Sint16
-                   -> IFloat
-                   -> Ref s1 ('Struct "spi_transaction_result")
-                   -> Ref s2 ('Struct "xyz")
-                   -> Ivory eff ()
-convert_sample_cal xoff yoff zoff scale res s = do
+convert_sample XyzCal {..} sens_scale res s = do
   f ((res ~> rx_buf) ! 1)
     ((res ~> rx_buf) ! 2)
     (s ~> XYZ.x)
-    xoff
+    cal_x_offset cal_x_scale
   f ((res ~> rx_buf) ! 3)
     ((res ~> rx_buf) ! 4)
     (s ~> XYZ.y)
-    yoff
+    cal_y_offset cal_y_scale
   f ((res ~> rx_buf) ! 5)
     ((res ~> rx_buf) ! 6)
     (s ~> XYZ.z)
-    zoff
+    cal_z_offset cal_z_scale
   where
-  f loref hiref resref offset = do
+  f loref hiref resref offset axis_scale = do
     lo <- deref loref
     hi <- deref hiref
     (u16 :: Uint16) <- assign ((safeCast lo) + ((safeCast hi) `iShiftL` 8))
     (i16 :: Sint16) <- assign (twosComplementCast u16)
-    (r :: IFloat)   <- assign (scale * safeCast (i16 + offset))
+-- From PX4: 
+-- float x_in_new = ((xraw_f * _accel_range_scale) - _accel_scale.x_offset) * _accel_scale.x_scale;
+--    (r :: IFloat)   <- assign (((safeCast i16 * sens_scale) - offset) * axis_scale)
```

Comment:
If we're using the qGroundControl calibration, we should be using this calculation of `r`, right?